### PR TITLE
Implement CSSKeyframesRule.name

### PR DESCRIPTION
**Pull request by canova**
Implementation of CSSKeyframesRule.name

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14576)

### CODE REVIEW

**Comment by canaltinova** [comment]
In spec there is a line:
> Setting this property to a value matching a CSS-wide keyword or any keyword defined for the animation-name property will throw a SyntaxError exception. 

But I tested this on firefox, chrome and safari, it seems they are not doing that. Should we follow the spec or do what major browsers do?
cc @Manishearth

**Comment by Manishearth** [comment]
Follow the spec, file bugs on other browsers (implement in Firefox!). Having keyframes that can't be referred to is a useless feature to have around, which we can just break.

Alternatively, follow the spec and file a spec bug. I suspect the resolution will be to fix other browsers.

**Comment by canaltinova** [comment]
Okay, I'm fixing this and I'll file a bug on other browsers :)

**Comment by canaltinova** [comment]
Updated the code to follow the spec.

**Comment by Manishearth** [comment]
Actually, this might not be a browser bug. The spec lets you define keyframe rules using strings that match existing idents (https://drafts.csswg.org/css-animations/#typedef-keyframes-name).

It should allow reserved idents here too.

**Comment by Manishearth** [comment]
https://github.com/w3c/csswg-drafts/issues/801

File: components/script/dom/csskeyframesrule.rs [hunk 1]
```diff
@@ -101,6 +103,23 @@ impl CSSKeyframesRuleMethods for CSSKeyframesRule {
...
+
+    // https://drafts.csswg.org/css-animations/#dom-csskeyframesrule-name
+    fn Name(&self) -> DOMString {
+        DOMString::from(&*self.keyframesrule.read().name)
+    }
+
+    // https://drafts.csswg.org/css-animations/#dom-csskeyframesrule-name
+    fn SetName(&self, value: DOMString) -> ErrorResult {
+        // Setting this property to a CSS-wide keyword or `none` will
+        // throw a Syntax Error.
```
Comment (Manishearth) [inline components/script/dom/csskeyframesrule.rs]: link to the spec bug I filed here

**Comment by canaltinova** [comment]
Oh, that makes sense. Undoing changes. Thanks

File: tests/wpt/web-platform-tests/cssom/CSSKeyframesRule.html [hunk 2]
```diff
@@ -40,18 +41,26 @@
...
+        assert_equals(keyframe.cssRules.length, 3, "CSSKeyframesRule cssRule length attribute after deleteRule function");
         assert_equals(keyframe.cssRules[0].cssText, "0% { top: 0px; }", "CSSKeyframesRule cssRule cssText attribute after deleteRule function");
         assert_equals(keyframe.cssRules[1].cssText, "50% { top: 100px; }", "CSSKeyframesRule cssRule cssText attribute after deleteRule function");
         assert_equals(keyframe.cssRules[2].cssText, "0% { top: 50px; }", "CSSKeyframesRule cssRule cssText attribute after deleteRule function");
         assert_equals(keyframe.cssRules[3], undefined, "CSSKeyframesRule cssRule cssText attribute after deleteRule function");
+
+        keyframe.name = "bar";
+        assert_equals(keyframe.name, "bar", "CSSKeyframesRule name setter");
+
+        keyframe.name = "initial";
```
Comment (Manishearth) [inline tests/wpt/web-platform-tests/cssom/CSSKeyframesRule.html]: You need to `assert_throws` here

**Comment by Manishearth** [comment]
r=me with spec bug link and using `assert_throws`

@bors-servo delegate+

File: components/script/dom/csskeyframesrule.rs [hunk 3]
```diff
@@ -101,6 +103,24 @@ impl CSSKeyframesRuleMethods for CSSKeyframesRule {
...
+    fn Name(&self) -> DOMString {
+        DOMString::from(&*self.keyframesrule.read().name)
+    }
+
+    // https://drafts.csswg.org/css-animations/#dom-csskeyframesrule-name
+    fn SetName(&self, value: DOMString) -> ErrorResult {
+        // https://github.com/w3c/csswg-drafts/issues/801
+        // Setting this property to a CSS-wide keyword or `none` will
+        // throw a Syntax Error.
+        if value == "initial" || value == "inherit" ||
```
Comment (emilio) [inline components/script/dom/csskeyframesrule.rs]: This should probably be ascii case-insensitive comparison probably?

But given not even Gecko does this, do we really care?

File: components/script/dom/csskeyframesrule.rs [hunk 4]
```diff
@@ -101,6 +103,27 @@ impl CSSKeyframesRuleMethods for CSSKeyframesRule {
...
+        // https://github.com/w3c/csswg-drafts/issues/801
+        // Setting this property to a CSS-wide keyword or `none` will
+        // throw a Syntax Error.
+        match_ignore_ascii_case! { value,
+            "initial" => return Err(Error::Syntax),
+            "inherit" => return Err(Error::Syntax),
+            "unset" => return Err(Error::Syntax),
+            "none" => return Err(Error::Syntax),
+            _ => ()
+        }
```
Comment (canaltinova) [inline components/script/dom/csskeyframesrule.rs]: @emilio You're right. I fixed it to ascii case-insensitive. But I couldn't write it like  `"initial" || inherit` in match_ignore_ascii_case. Is there a way to reduce the code duplication?

Comment (emilio) [inline components/script/dom/csskeyframesrule.rs]: I don't think so :(

**Pull request merged**